docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,281 @@
+
+This pattern is used in WASM processors. A server (in this case: Conduit)
+listens to commands (in this case: via a bidirectional gRPC stream). A client (
+in this case: a connector) sends a command to either register a schema or fetch
+a schema. Conduit receives the command and replies. An example can be seen below:
+
+```protobuf
+rpc CommandStream(stream Command) returns (stream Response);
+```
+
+For different types of commands and response to be supported, `Command`
+and `Response` need to have a `oneof` field where all the possible commands
+and respective responses are listed:
+
+```protobuf
+message Command {
+    oneof cmd {
+        SaveSchemaCommand saveSchemaCmd = 1;
+        // etc.
+   }
+}
+
+message Response {
+  oneof resp {
+    SaveSchemaResponse saveSchemaResp = 1;
+    // etc.
+  }
+}
+```    
+**Advantages**:
+
+1. No additional connection setup. When Conduit starts a connector process, it
+   establishes a connection. The same connection is used for all communication (
+   e.g. configuring a connector, opening, reading/writing records, etc.)
+2. Connector actions (which are planned for a future milestone) might use the
+   same command-and-reply stream.
+
+**Disadvantages**:
+
+1. Separate flow needed to establish a connector to a remote Conduit instance (
+   see [requirement](#requirements) #3).
+2. A single method for all the operations makes both, the server and client
+   implementation, more complex. In Conduit, a single gRPC method needs to check
+   for the command type and then reply with a response. Then the client (i.e.
+   the connector) needs to check the response type. In case multiple commands
+   are sent, we need ordering guarantees.
+
+#### Option 2: Exposing Conduit as a service
+
+Conduit exposes a service to work with schemas. Connectors access the service
+and call methods on the service. 
+
+For this work, a connector (i.e. clients of the schema service) needs Conduit's
+IP address and the gRPC port. The IP address can be fetched
+using [peer](https://pkg.go.dev/google.golang.org/grpc/peer#Peer). Conduit can
+send its gRPC port to the connector via the `Configure` method.
+
+**Advantages**: 
+
+1. Works with a remote Conduit instance (see [requirement](#requirements) #3).
+2. Easy to understand: the gRPC methods, together with requests and responses,
+   can easily be understood from a proto file.
+3. An HTTP API for the schema registry can easily be exposed (if needed).
+
+**Disadvatanges**:
+
+1. Changes needed to communicate Conduit's gRPC port to the connector.
+
+#### Chosen option
+
+**Option 2** is the chosen method since it offers more clarity and the support for
+remote Conduit instances.
+
+## Questions
+
+1. **Q**: Should a schema be accessible across Conduit instances?
+
+   Schemas need to be accessible across pipelines (requirement #3). If the
+   schema registry is part of Conduit, should we make it accessible via the API?
+   
+   **A**:
```

Comment:
Running one pipeline per Conduit instance could be a thing, so it looks like the answer to this is "yes".